pennylane/devices/qutrit_mixed/measure.py

```diff
@@ -0,0 +1,309 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code relevant for performing measurements on a qutrit mixed state.
+"""
+
+from typing import Callable
+from string import ascii_letters as alphabet
+from pennylane import math
+from pennylane.ops import Sum, Hamiltonian
+from pennylane.measurements import (
+    StateMeasurement,
+    MeasurementProcess,
+    ExpectationMP,
+    StateMP,
+    ProbabilityMP,
+    VarianceMP,
+)
+from pennylane.operation import Observable
+from pennylane.typing import TensorLike
+
+from .utils import (
+    get_einsum_mapping,
+    resquare_state,
+    get_num_wires,
+    get_new_state_einsum_indices,
+    QUDIT_DIM,
+)
+from .apply_operation import apply_operation
+
+alphabet_array = math.asarray(list(alphabet))
+
+
+def apply_observable_einsum(obs: Observable, state, is_state_batched: bool = False):
+    r"""Applies an observable to a density matrix rho, giving obs@state
+
+    Args:
+        obs (Operator): Operator to apply to the quantum state
+        state (array[complex]): Input quantum state
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+
+    Returns:
+        TensorLike: the result of obs@state
+    """
+
+    def map_indices(
+        state_indices, row_indices, new_row_indices, **kwargs
+    ):  # pylint: disable=unused-argument
+        """Map indices to wires
+        Args:
+            state_indices (str): Indices that are summed
+            row_indices (str): Indices that must be replaced with sums
+            new_row_indices (str): Tensor indices of the state
+            **kwargs (dict): Stores indices calculated in `get_einsum_mapping`
+
+        Returns:
```

Comment:
Mind adding a section like Args for the closure variables? I'm having a hard time figuring out which variables from the outer scope we are relying on.

```
Closure:
    val1
```

If we are not relying on any variables from outer scope, could we pull this out to be a global function? That can make things a lot easier to debug and follow.

Reply:
I am going to move it out. I will add all the expected kwargs to docstring too.